SignalServiceKit/src/Storage/OWSMediaGalleryFinder.m

```diff
@@ -0,0 +1,142 @@
+}
+
+- (NSUInteger)mediaIndexForMessage:(TSMessage *)message transaction:(YapDatabaseReadTransaction *)transaction
+{
+    NSString *groupId;
+    NSUInteger index;
+    
+    [[self galleryExtensionWithTransaction:transaction] getGroup:&groupId index:&index forKey:message.uniqueId inCollection:[TSMessage collection]];
+    
+    return index;
+}
+
+- (void)enumerateMediaMessagesWithThread:(TSThread *)thread
+                             transaction:(YapDatabaseReadTransaction *)transaction
+                                   block:(void (^)(TSMessage *))messageBlock
+{
+    NSString *group = [self mediaGroupWithThreadId:thread.uniqueId];
+    [[self galleryExtensionWithTransaction:transaction]
+        enumerateKeysAndObjectsInGroup:group
+                            usingBlock:^(NSString *_Nonnull collection,
+                                NSString *_Nonnull key,
+                                id _Nonnull object,
+                                NSUInteger index,
+                                BOOL *_Nonnull stop) {
+                                OWSAssert([object isKindOfClass:[TSMessage class]]);
+                                messageBlock((TSMessage *)object);
+                            }];
+}
+
+#pragma mark - Util
+
+- (YapDatabaseAutoViewTransaction *)galleryExtensionWithTransaction:(YapDatabaseReadTransaction *)transaction
+{
+    YapDatabaseAutoViewTransaction *extension = [transaction extension:OWSMediaGalleryFinderExtensionName];
+    OWSAssert(extension);
+    
+    return extension;
+}
+
++ (NSString *)mediaGroupWithThreadId:(NSString *)threadId
+{
+    return [NSString stringWithFormat:@"%@-media", threadId];
+}
+
+- (NSString *)mediaGroupWithThreadId:(NSString *)threadId
+{
+    return [[self class] mediaGroupWithThreadId:threadId];
+}
+
+#pragma mark - Extension registration
+
++ (void)asyncRegisterDatabaseExtensionsWithPrimaryStorage:(OWSStorage *)storage
+{
+    [storage asyncRegisterExtension:[self mediaGalleryDatabaseExtension]
+                           withName:OWSMediaGalleryFinderExtensionName];
+}
+
++ (YapDatabaseAutoView *)mediaGalleryDatabaseExtension
+{
+    YapDatabaseViewSorting *sorting = [YapDatabaseViewSorting withObjectBlock:^NSComparisonResult(YapDatabaseReadTransaction * _Nonnull transaction, NSString * _Nonnull group, NSString * _Nonnull collection1, NSString * _Nonnull key1, id  _Nonnull object1, NSString * _Nonnull collection2, NSString * _Nonnull key2, id  _Nonnull object2) {
+        
+        if (![object1 isKindOfClass:[TSMessage class]]) {
+            return NSOrderedSame;
+        }
+        TSMessage *message1 = (TSMessage *)object1;
+        
+        if (![object2 isKindOfClass:[TSMessage class]]) {
+            return NSOrderedSame;
+        }
+        TSMessage *message2 = (TSMessage *)object2;
+        
+        return [@(message1.timestampForSorting) compare:@(message2.timestampForSorting)];
```

Comment:
I wonder if we could pull this out into a NSObject+OWS category: `[NSObject orderingForObject:(NSObject *)left andObject:(NSObject *)right expectedClass:(Class)clazz comparisonBlock:(id(^_NonNull)(id))comparisonBlock]`.

then:
```

+ (YapDatabaseAutoView *)mediaGalleryDatabaseExtension
{
    YapDatabaseViewSorting *sorting = [YapDatabaseViewSorting withObjectBlock:^NSComparisonResult(YapDatabaseReadTransaction * transaction, NSString * group, NSString * collection1, NSString * key1, id  object1, NSString * collection2, NSString * key2, id  object2) {
return [NSObject orderingForObject:message1 andObject:message2 expectedClass:[TSMessage class] comparisonBlock:^(TSMessage message) { return message. timestampForSorting; }];
}
```

Reply:
TIOLI obvs

Reply:
Though we could use `expectedClass` to dynamically check, we'd still have to cast in the comparisonBlock to avoid compiler warnings right? (unless we wrote it in Swift)